### DISCUSSION TITLE
Submodule path error for installing Task_1 code

### DISCUSSION THREAD

**Discussion by jingyunc** [FETS-AI/Challenge#190] [Q&A]
I was try to install Task 1 code for the FeTS24 Challenge. I followed the [Task 1 instructions](https://github.com/FeTS-AI/Challenge/blob/main/Task_1/README.md) until step 8 (pip install .), where I got following error:

```
  Preparing metadata (setup.py) ... done
Collecting openfl@ git+https://github.com/intel/openfl.git@f4b28d710e2be31cdfa7487fdb4e8cb3a1387a5f (from fets-challenge==2.0)
  Cloning https://github.com/intel/openfl.git (to revision f4b28d710e2be31cdfa7487fdb4e8cb3a1387a5f) to /tmp/pip-install-umyuqv17/openfl_6de81b4493e04a72a7842469596cdbd7
  Running command git clone --filter=blob:none --quiet https://github.com/intel/openfl.git /tmp/pip-install-umyuqv17/openfl_6de81b4493e04a72a7842469596cdbd7
  Running command git rev-parse -q --verify 'sha^f4b28d710e2be31cdfa7487fdb4e8cb3a1387a5f'
  Running command git fetch -q https://github.com/intel/openfl.git f4b28d710e2be31cdfa7487fdb4e8cb3a1387a5f
  Running command git checkout -q f4b28d710e2be31cdfa7487fdb4e8cb3a1387a5f
  Resolved https://github.com/intel/openfl.git to commit f4b28d710e2be31cdfa7487fdb4e8cb3a1387a5f
  Preparing metadata (setup.py) ... done
Collecting GANDLF@ git+https://github.com/CBICA/GaNDLF.git@e4d0d4bfdf4076130817001a98dfb90189956278 (from fets-challenge==2.0)
  Cloning https://github.com/CBICA/GaNDLF.git (to revision e4d0d4bfdf4076130817001a98dfb90189956278) to /tmp/pip-install-umyuqv17/gandlf_7bee759db7a34001b96f1fd3e9852230
  Running command git clone --filter=blob:none --quiet https://github.com/CBICA/GaNDLF.git /tmp/pip-install-umyuqv17/gandlf_7bee759db7a34001b96f1fd3e9852230
  Running command git rev-parse -q --verify 'sha^e4d0d4bfdf4076130817001a98dfb90189956278'
  Running command git fetch -q https://github.com/CBICA/GaNDLF.git e4d0d4bfdf4076130817001a98dfb90189956278
  Running command git checkout -q e4d0d4bfdf4076130817001a98dfb90189956278
  Resolved https://github.com/CBICA/GaNDLF.git to commit e4d0d4bfdf4076130817001a98dfb90189956278
  Running command git submodule update --init --recursive -q
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Installing backend dependencies ... done
  Preparing metadata (pyproject.toml) ... done
Collecting fets@ git+https://github.com/FETS-AI/Algorithms.git@fets_challenge (from fets-challenge==2.0)
  Cloning https://github.com/FETS-AI/Algorithms.git (to revision fets_challenge) to /tmp/pip-install-umyuqv17/fets_df05ec67c4384f9e8c1a1a3531332640
  Running command git clone --filter=blob:none --quiet https://github.com/FETS-AI/Algorithms.git /tmp/pip-install-umyuqv17/fets_df05ec67c4384f9e8c1a1a3531332640
  Running command git checkout -b fets_challenge --track origin/fets_challenge
  Switched to a new branch 'fets_challenge'
  Branch 'fets_challenge' set up to track remote branch 'fets_challenge' from 'origin'.
  Resolved https://github.com/FETS-AI/Algorithms.git to commit 60e0b8761229edde18e3d707e3e3e5eb0c0fb80f
  Running command git submodule update --init --recursive -q
  **fatal: No url found for submodule path 'GANDLF' in .gitmodules**
  error: subprocess-exited-with-error
  
  × git submodule update --init --recursive -q did not run successfully.
  │ exit code: 128
  ╰─> See above for output.
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
error: subprocess-exited-with-error

× git submodule update --init --recursive -q did not run successfully.
│ exit code: 128
╰─> See above for output.
```

Can anyone please suggest a fix? Thanks.

**Comment by sunyeongan** [comment]
https://github.com/FeTS-AI/Challenge/blob/524d6b9f489ba852d601a437839abd937e5cab89/Task_1/setup.py#L32

In my opinion, the problem appears to be caused by GANDALF's URL not existing. This needs to be fixed.

**Reply by sunyeongan** [reply]
Try running this command in terminal
`pip install git+https://github.com/CBICA/GaNDLF.git@e4d0d4bfdf4076130817001a98dfb90189956278`

**Reply by sunyeongan** [reply]
Check out this issue

https://github.com/FeTS-AI/Challenge/issues/172

**Reply by jingyunc** [reply]
Thanks @sunyeongan for looking into this. Unfortunately, running `pip install git+https://github.com/CBICA/GaNDLF.git@e4d0d4bfdf4076130817001a98dfb90189956278` resulted in a series of package requirement conflicts.

Following Geeks-Sid's steps in #172, I was able to install gandlf. However, when I reran "pip install ." or just "pip install git+https://github.com/FETS-AI/Algorithms.git@fets_challenge", I got the same error as posted above. The error seemed from fets package.

**Reply by sunyeongan** [reply]
@jingyunc hello 
I also experienced a similar issue.
The reason why the SimpleITK version does not match, the versions of various libraries do not match, so I forked two GaNDLF and FeTS-Algorithms repositories in my repository and modified the versions one by one, then changed line 32 of setup.py.

**Reply by sarthakpati** [reply]
@sunyeongan - could you provide the specifics of which repo/setup line you changed? Thanks!

**Reply by sunyeongan** [reply]
Hi @sarthakpati 


At first, I modified the URL of install_requires in setup.py, https://github.com/FeTS-AI/Challenge/blob/524d6b9f489ba852d601a437839abd937e5cab89/Task_1/setup.py#L31
but an error occurred. So, I git cloned each library openfl, GaNDLF, and fets and installed them individually using pip. The process I did is below.

**openfl**
`git clone https://github.com/intel/openfl.git`
`cd openfl`
`git checkout f4b28d710e2be31cdfa7487fdb4e8cb3a1387a5f`
`pip install .`


**GaNDLF**
`git clone https://github.com/CBICA/GaNDLF.git`
`cd GaNDLF`
`git checkout e4d0d4bfdf4076130817001a98dfb90189956278`
`pip install .`

**fets**
`git clone https://github.com/FETS-AI/Algorithms.git`
`cd Algorithms`
`git checkout 60e0b8761229edde18e3d707e3e3e5eb0c0fb80f`
`pip install .`

A discussion related to this was carried out in the issue below.
https://github.com/FeTS-AI/Challenge/issues/188#issuecomment-2150749855

I think 
https://github.com/FeTS-AI/Challenge/blob/524d6b9f489ba852d601a437839abd937e5cab89/Task_1/setup.py#L32
should be changed to `GANDLF @ git+https://github.com/mlcommons/GaNDLF.git@e4d0d4bfdf4076130817001a98dfb90189956278`

**Reply by sunyeongan** [reply]
When checking each library with the `pip show` command, if the version is as shown in the picture below, the FeTS_Challenge.py code runs normally.

**openfl**
- version : 1.3

**GaNDLF**
- version : 0.0.14.dev0


<img width="1562" alt="image" src="https://github.com/FeTS-AI/Challenge/assets/44018024/7430a2fd-7c53-4c9f-94b6-6997ca333600">
